nixos/modules/programs/steam.nix

```diff
@@ -4,52 +4,114 @@ with lib;
+    (cfg.gamescopeSession.enable && gamescopeCfg.capSysNice)
+    {
+      buildFHSUserEnv = pkgs.buildFHSUserEnvBubblewrap.override {
+        # use the setuid wrapped bubblewrap
+        bubblewrap = "/run/wrappers";
+      };
+    }
+  );
+
+  gamescopeSessionFile = (pkgs.writeTextDir "share/wayland-sessions/steam.desktop" ''
+    [Desktop Entry]
+    Name=Steam
+    Comment=A digital distribution platform
+    Exec=${pkgs.writeShellScript "steam-gamescope" ''
+      ${let
+        exports = builtins.attrValues (builtins.mapAttrs (n: v: "export ${n}=${v}") cfg.gamescopeSession.env);
+      in
+        builtins.concatStringsSep "\n" exports}
+        gamescope --steam ${toString cfg.gamescopeSession.args} -- steam -tenfoot -pipewire-dmabuf
+    ''}
+    Type=Application
+  '').overrideAttrs (_: { passthru.providedSessions = [ "steam" ]; });
+in
+{
   options.programs.steam = {
-    enable = mkEnableOption (lib.mdDoc "steam");
+    enable = mkEnableOption (mdDoc "steam");
 
     remotePlay.openFirewall = mkOption {
       type = types.bool;
       default = false;
-      description = lib.mdDoc ''
+      description = mdDoc ''
         Open ports in the firewall for Steam Remote Play.
       '';
     };
 
     dedicatedServer.openFirewall = mkOption {
       type = types.bool;
       default = false;
-      description = lib.mdDoc ''
+      description = mdDoc ''
         Open ports in the firewall for Source Dedicated Server.
       '';
     };
+
+    gamescopeSession = mkOption {
+      description = mdDoc "Run a GameScope driven Steam session from your display-manager";
+      type = types.submodule {
+        options = {
+          enable = mkEnableOption (mdDoc "GameScope Session");
+          args = mkOption {
+            type = types.listOf types.string;
+            default = [ ];
+            description = mdDoc ''
+              Arguments to be passed to GameScope for the session.
+            '';
+          };
+
+          env = mkOption {
+            type = types.attrsOf types.string;
+            default = { };
+            description = mdDoc ''
+              Environmental variables to be passed to GameScope for the session.
+            '';
+          };
+        };
+      };
+    };
   };
 
   config = mkIf cfg.enable {
-    hardware.opengl = { # this fixes the "glXChooseVisual failed" bug, context: https://github.com/NixOS/nixpkgs/issues/47932
+    hardware.opengl = {
+      # this fixes the "glXChooseVisual failed" bug, context: https://github.com/NixOS/nixpkgs/issues/47932
       enable = true;
       driSupport = true;
       driSupport32Bit = true;
     };
 
+    security.wrappers = mkIf (cfg.gamescopeSession.enable && gamescopeCfg.capSysNice) {
+      # needed or steam fails
+      bwrap = {
+        owner = "root";
+        group = "root";
+        source = "${pkgs.bubblewrap}/bin/bwrap";
+        setuid = true;
+      };
+    };
+
+    programs.gamescope.enable = mkIf cfg.gamescopeSession.enable true;
```

Comment:
```suggestion
    programs.gamescope.enable = mkDefault cfg.gamescopeSession.enable;
```

Reply:
I opted for:
```nix
{
    programs.gamescope.enable = mkIf cfg.gamescopeSession.enable (mkDefault true);
}
```

Reply:
Nit: I find it a bit weird to have true behind a condition. Might as well just have the condition in there directly.

Reply:
yeah you are right, I must not have been thinking correctly :sweat_smile: